MAINT: deal with cases in `minimize` for `(bounds.lb == bounds.ub).any() (third alternate for #12889)

andyfaff opened the pull request:
@mdhaber, how about this factoring out?

mdhaber commented:
Before I look in detail, what is better than gh-13096? Could gh-13096 be improved?

andyfaff commented:
@mdhaber, this approach:

- puts the factorisation down into the various `_minimize_*`, which means that `minimize` and `fmin_*` should work without alteration.
- #13096 would have had difficulties with the constraints functions for SLSQP, because they all need to be wrapped as well, and it'd be a pain to do that in `minimize`.
- `eps` and `finite_diff_rel_step` can also be arrays, which also need to be reduced.

andyfaff commented:
I think the only issue as it stands is how to re-expand the `LbfgsInvHessProduct`, I don't know how to do that.

andyfaff commented on scipy/optimize/slsqp.py:
```diff
@@ -284,6 +312,9 @@ def _minimize_slsqp(func, x0, args=(), jac=None, bounds=None,
             raise ValueError('Constraint %d has no function defined.' % ic)
+        cfun = con['fun']
+        if equal_bounds.any():
+            cfun = _equal_bounds_func(cfun, full_x0, equal_bounds)
```
@mdhaber, it's all these that one has to wrap that are the pain in the neck.

andyfaff commented:
Decided to take out the changes to `_numdiff` and `_differentiable_functions` because they muddy the water a little and aren't critical to this PR.

andyfaff commented:
@nmayorov, @larsoner, this is an alternate way to fix the equal bounds bug. This alternate version factors out the parameters which have `lb==ub`. It only does this for SLSQP and LBFGSB, as they are the only two that display problems in the test suite.

andyfaff commented:
When I get to the `OptimizeResult` construction `x`, and `jac` need to be expanded. It's relatively easy to do, but should I use `np.nan`, or `0`, for those entries in `jac` that have equal lower and upper bounds?

andyfaff commented:
Rebased and squashed commits.

@rgommers, it would be good if a solution to the original issue is in the next release (there have been several issues raised over this). I've made different approaches to a solution and @mdhaber did as well (#13096). From my point of view I think this PR is a better solution to the issue than Matt's, for the reasons outlined above.

mdhaber commented on scipy/optimize/lbfgsb.py:
```diff
@@ -268,42 +269,80 @@ def _minimize_lbfgsb(fun, x0, args=(), jac=None, bounds=None,
     arrive at `ftol`.
-
+    If any of the lower bounds are equal to the upper bounds then the
+    corresponding entry in ``OptimizeResult.jac`` will be np.nan.
```
I think `nan` is the appropriate choice. After all, I think that the main thing we didn't like about the other approach is that there was no basis for setting the derivatives to zero.

mdhaber commented on scipy/optimize/slsqp.py:
```diff
@@ -440,7 +474,9 @@ def cjac(x, *args):
-                callback(np.copy(x))
+                new_x = np.copy(full_x0)
+                new_x[~filtered_params] = x
+                callback(new_x)
```
Why not use `_filter_params_func` on `callback` when it is passed in?

andyfaff replied:
Great idea.

mdhaber commented on scipy/optimize/lbfgsb.py:
```diff
@@ -268,42 +269,80 @@ def _minimize_lbfgsb(fun, x0, args=(), jac=None, bounds=None,
+
+    # initial vector must lie within the bounds. Otherwise ScalarFunction and
+    # approx_derivative will cause problems
+    full_x0 = np.clip(full_x0, full_new_bounds[0], full_new_bounds[1])
```
Was this an existing bug that you are fixing here?

andyfaff replied:
It's [existing behaviour](https://github.com/scipy/scipy/blob/master/scipy/optimize/lbfgsb.py#L298). When a `ScalarFunction` is created it evaluates `fun` and `grad` at an initial point. If that initial point is outside the bounds then a finite difference estimation of grad raises an Exception.

mdhaber replied:
I see it's just moved from below

mdhaber commented on scipy/optimize/lbfgsb.py:
```diff
@@ -268,42 +269,80 @@ def _minimize_lbfgsb(fun, x0, args=(), jac=None, bounds=None,
+    x = full_x0[~filtered_params]
+    n, = x.shape
+
+    # if there are equal bounds then the user function has to be wrapped.
```
```suggestion
    # if there are filtered parameters then the user function has to be wrapped.
```
Because you can have equal bounds and not do the filtering, as you note above.

mdhaber commented on scipy/optimize/lbfgsb.py:
```diff
@@ -37,7 +37,8 @@
 from .optimize import (MemoizeJac, OptimizeResult,
-                       _check_unknown_options, _prepare_scalar_function)
+                       _check_unknown_options, _prepare_scalar_function,
+                       _filter_params_func, _filter_params_jac)
```
```suggestion
                       _filter_params_func)
```
Not actually used.

mdhaber commented on scipy/optimize/lbfgsb.py:
```diff
@@ -268,42 +269,80 @@ def _minimize_lbfgsb(fun, x0, args=(), jac=None, bounds=None,
     if bounds is None:
-        bounds = [(None, None)] * n
-    if len(bounds) != n:
+        bounds = [(-np.inf, np.inf)] * full_x0.size
```
I didn't notice in a first pass why this was needed. Why convert to `inf` here then back to `None`?

andyfaff replied:
All the preparation work (`np.clip`, `approx_derivative` (via `ScalarFunction`), `full_new_bounds[0] == full_new_bounds[1]`) requires comparison of numerical arrays. This is why `inf` is used here.

However, LBFGSB Fortran code uses `None` to represent an unbounded lower/upper limit, so any non-finite bounds have to be converted to `None` later on.

mdhaber commented on scipy/optimize/lbfgsb.py:
```diff
@@ -393,10 +434,26 @@ def _minimize_lbfgsb(fun, x0, args=(), jac=None, bounds=None,
+        y_amended = np.zeros_like(s_amended)
+        s_amended[:, ~filtered_params] = hess_inv.sk
+        y_amended[:, ~filtered_params] = hess_inv.yk
+        hess_inv = LbfgsInvHessProduct(s_amended, y_amended)
```
I'll need to take a closer look at this later. In the meantime - do you need to create a `LbfgsInvHessProduct` above and use it to create this one, or could you somehow create only this one? Can/should entries be `nan` instead of 0?

andyfaff replied:
If these modifications aren't done then `hess_inv.todense()` results in an array that has shape `(np.count_nonzero(filtered_params), np.count_nonzero(filtered_params))`, i.e. the size of the reduced problem. This might be confusing to people that expect it to have a size corresponding to `full_x0`.

I tried to set `s_amended` to 0 for filtered parameters, and `y_amended` to `np.nan`. However, when one calculates `hess_inv.todense()` all the entries come out as `np.nan`, which is not correct either.

So, I'm not sure which behaviour is the worst, or if there is another way of achieving this?

mdhaber replied:
> If these modifications aren't done

I know you have to modify it. I just mean can you modify it by changing the inputs to the original call to the initializer (`LbfgsInvHessProduct`) rather than calling the initializer once and using attributes of the object to call it again.

Then again, I'm wondering whether we can salvage this at all. This is the _inverse_ of the Hessian, not the Hessian itself. Suppose we were trying to comput this directly (rather than approximate it as is done in LBFGSB, right?) When we don't take the second partials with respect to the fixed variables,  we would have to `nan` out some rows and columns corresponding to the fixed variables. What would happen when we took the inverse?

andyfaff replied:
It's possible to create the inverse hessian with different inputs (i.e. with nan for grad entries, and 0 for sk entries). I don't know how to judge if this creates a valid LbfgsInvHessProduct or not.

mdhaber commented on scipy/optimize/lbfgsb.py:
```diff
@@ -268,42 +269,80 @@ def _minimize_lbfgsb(fun, x0, args=(), jac=None, bounds=None,
+    else:
+        filtered_params = full_new_bounds[0] == full_new_bounds[1]
+
+    x = full_x0[~filtered_params]
```
Why not call it `x0` and avoid some changes below?

andyfaff replied:
I thought it might be good to keep `x0` as an unchanged array of what the user passed in to the function.

mdhaber commented on scipy/optimize/lbfgsb.py:
```diff
@@ -362,7 +401,9 @@ def _minimize_lbfgsb(fun, x0, args=(), jac=None, bounds=None,
-                callback(np.copy(x))
+                new_x = np.copy(full_x0)
+                new_x[~filtered_params] = x
+                callback(new_x)
```
Can you wrap `callback` with `_filter_params_func`?

mdhaber commented on scipy/optimize/optimize.py:
```diff
@@ -287,6 +287,24 @@ def _check_clip_x(x, bounds):
+    return wrapped_func
+
+
+def _filter_params_jac(jac, mask):
```
Even though the parameter that specifies the gradient function is called `jac`, I think we should use the term `grad` for the partials of a scalar objective function and reserve `jac` for the partials of a vector-valued function.

On a related note, this doesn't yet work for the jacobian of vector-valued constraints.
```
import numpy as np
import scipy.optimize as optimize

def f(x):
    return (x**3).sum()

def con(x):
    return x**2-1

def conjac(x):
    return np.diag(2*x)

def callback(x):
    print(x.shape)

method = 'slsqp'
constraints={"fun":con, "type":'ineq', "jac":conjac}
x0 = np.array([0.5, 3.0, 2, 5])
bounds = [(0.0, 3.0), (2.0, 2), (0, np.inf), (1.5, 1.5)]

res = optimize.minimize(
    f, x0, method=method, bounds=bounds, constraints=constraints, callback=callback
)
```

andyfaff replied:
SLSQP constraints should be scalar functions, `R^n --> 1`. You can have more than one constraint, but each one needs to be scalar.

andyfaff replied:
jac and grad are littered around the codebase all over the place, I get your point.
Fixed the vector valued constraint problem and added a test.

mdhaber commented on scipy/optimize/slsqp.py:
```diff
@@ -249,17 +254,50 @@ def _minimize_slsqp(func, x0, args=(), jac=None, bounds=None,
+    # This definitely needs to be done when jacobians are estimated
+    # numerically. However, even if the user supplies a callable jacobian
+    # filtering is advised because some of the user supplied constraints may
+    # not have callable jacobians.
```
Could check for that, right?

andyfaff replied:
We could. However, it would require a subsequent change to filter everything (including a double pass through the constraints loop) if one of them didn't have a callable jacobian. I reckon this would be quite complicated.

- I think whilst equal bounds occur, they're not the majority of problems (say 1%).
- If equal bounds do occur there probably won't be callable jacobians for all constraints and the original function (say 10%).

Since the overall probability was so low, I didn't think it'd be worth implementing for the complexity involved.

mdhaber commented on scipy/optimize/slsqp.py:
```diff
@@ -249,17 +254,50 @@ def _minimize_slsqp(func, x0, args=(), jac=None, bounds=None,
 
     # Transform x0 into an array.
-    x = asfarray(x0).flatten()
+    full_x0 = np.copy(asfarray(x0).flatten())
```
A lot of this is repeated from above, and it may have to be done for `trust-constr` at some point. Can the code be shared?

andyfaff replied:
When the changes were made that caused the issue for SLSQP and LBFGSB they didn't affect `trust-constr`, the behaviour of `trust-constr` w.r.t equal bounds has been unchanged since its inception.
`trust-constr` is quite complicated, with a lot more user specifiable options and output. I think the approach may be able to be shared, but I'm not sure about code.
In this PR I'm not sure if I could find any more code in the changes to SLSQP and LBFGSB that could be factored out into shared functions.

mdhaber commented on scipy/optimize/slsqp.py:
```diff
@@ -346,30 +393,17 @@ def cjac(x, *args):
     jw = zeros(len_jw)
 
     # Decompose bounds into xl and xu
-    if bounds is None or len(bounds) == 0:
```
The diff is a bit confusing here, and being close to the end of the file my mind is starting to check out. What is all this for?

andyfaff replied:
SLSQP accepts two arrays for bounds, `xl` and `xu`. Instead of `-np.inf/np.inf` for no bounds on a given parameter it uses `np.nan`. The changes here are a rationalisation of the previous code. `full_new_bounds` (created at the top) already contains float arrays that can be used for `xl` and `xu` (the only thing that needed to be done was to convert any `np.inf` to `np.nan`).

mdhaber commented:
After reviewing, some thoughts in response to these. (They're not meant to be defensive about gh-13096, because that was really was just supposed to explore the issue. I really just want the least painful resolution, wherever that comes from.)

> puts the factorisation down into the various _minimize_*, which means that minimize and fmin_* should work without alteration.

This is actually a superficial difference.  In gh-13096, there is a big block of code _added_ at the beginning of `minimize` and a small block at the end of `minimize` that do all the heavy lifting. There are basically no changes or deletions. Each block could be put in a function, and these functions could instead be called at the beginning/end of e.g. `lbfgsb` and `slsqp`. I chose to ignore the legacy functions, but if we care about them, I think it would be simple to do this.

> #13096 would have had difficulties with the constraints functions for SLSQP, because they all need to be wrapped as well, and it'd be a pain to do that in minimize (there is a whole set of checking that needs to be done first, etc).

I'm not sure I understand. I had already dealt with constraint functions for SLSQP and their jacobian functions. This PR doesn't yet address jacobians for vector-valued constraints.

> eps and finite_diff_rel_step can also be arrays, which also need to be reduced.
Yup, I forgot about that.  I also hadn't done the modifications to `LbfgsInvHessProduct` yet; that's another advantage here.

I think we should go back and forth between the PRs, as there are advantages and disadvantages in each, and hopefully we will converge to something we can agree on. Does that sound good? So I'd propose that the next step would be for me to make some changes in gh-13096 to incorporate the advantages here. I think it's worth doing that because one advantage of gh-13096 is that it is non-invasive - it involves some additions, but they are very localized and there are basically no changes or deletions. What do you think?

andyfaff commented:
> I'm not sure I understand. I had already dealt with constraint functions for SLSQP and their jacobian functions. This PR doesn't yet address jacobians for vector-valued constraints.

Perhaps a first step would be adding the modifications I've done to `test_slsqp.py` to check that #13096 works with constraints (scalar and vector valued).

andyfaff commented:
> eps and finite_diff_rel_step can also be arrays, which also need to be reduced

The last commit added a test for that. The test fails if the `finite_diff_rel_step` array isn't filtered.

tylerjereddy commented:
I'm going to bump the milestone on this matter--there seems to be some pretty detailed discussion above. If this really is needed for 1.6.x series I suppose you might still have a little time or you could put in during the rc phase, but I can't really tell what the timeline for merging is and I'm going to try to keep moving forward with release process.

tylerjereddy commented:
I don't see any activity since I bumped the milestone  on the previous release branching. Furthermore, there still seem to be "competing PRs" without much clarity as to which one is to be preferred.

andyfaff commented:
I think the underlying issue is one of the biggest pain points for optimize. I think both @mdhaber and my PR are ok, but not optimal.

tylerjereddy commented:
Oh dear, I'll bump milestone--maybe when Matt switches back to `optimize` from `stats` focus for a bit the decision on this one might move forward a bit.
